### PR TITLE
FCN example deadlock fixes

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description
Currently the FCN example sometimes runs into a deadlock on multi-node systems. This PR tries to fix that by making the `samples_per_year` an exact multiple of the `world_size`. It also makes the `num_workers` for the dataloaders configurable in the training recipe. 

Closes: #312 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/datapipes/climate/era5_hdf5.py [hunk 1]
```diff
@@ -157,7 +157,9 @@ def parse_dataset_files(self) -> None:
         self.logger.info(f"Getting file stats from {self.data_paths[0]}")
         with h5py.File(self.data_paths[0], "r") as f:
             # truncate the dataset to avoid out-of-range sampling
-            data_samples_per_year = f["fields"].shape[0] - self.num_steps * self.stride
+            data_samples_per_year = (
```
Comment (Alexey-Kamenev) [inline modulus/datapipes/climate/era5_hdf5.py:160]: So the idea here is that each rank should have the same number of samples? Maybe add a comment to make it more clear?

**Review by Alexey-Kamenev — APPROVED** [review]

File: modulus/datapipes/climate/era5_hdf5.py (hunk 1 above)
Comment (ktangsali) [reply modulus/datapipes/climate/era5_hdf5.py:160]: That is correct! Good point about adding a comment. Let me do that

**Review by loliverhennigh — APPROVED** [review]

**Comment by stefan-maxar** [comment]
Perhaps I am missing something, but I think this logic might still deadlock since `data_samples_per_year` is only used if `self.num_samples_per_year` is `None` (line 175 in era5_hdf5.py) and given that `num_samples_per_year_train` is currently defined in the FCN config YAML? 

I think it the same logic as provided to `data_samples_per_year` can be applied to the config-defined `num_samples_per_year_train`...but am not quite sure. That being said, considering the yearly HDF5 files have a full sample of 1460 (365 days * 4 samples per day), perhaps there is a better way to do this such that all samples can be used and/or a fixed max number of samples based on world_size.

**Pull request merged**